### ISSUE TITLE
Cannot write to EXT3 formated external HDD

### ISSUE THREAD

**Issue by marcosfrm** [storaged-project/udisks#261]
(or any other native Linux filesystem)

I have one of these cheap SATA to USB converters:

https://ae01.alicdn.com/kf/HTB1T1W0HVXXXXcLXVXXq6xXFXXXV/-font-b-USB-b-font-3-0-To-Laptop-22P-2-5-font-b-SSD.jpg

The hard disk is unpartitioned and contains a fresh new EXT3 filesystem inside (in the whole device).

On up-to-date Arch (udisks 2.6.4, GNOME 3.24), when I mount it via Nautilus, my user does not have write permissions in the mount point.

```
$ ls -l /run/media/
total 0
drwxr-x---+ 3 root root 60 May 15 14:03 marcos

$ getfacl /run/media/marcos/
getfacl: Removing leading '/' from absolute path names
# file: run/media/marcos/
# owner: root
# group: root
user::rwx
user:marcos:r-x
group::---
mask::r-x
other::---

$ ls -l /run/media/marcos/
total 4
drwxr-xr-x 3 root root 4096 May 15 14:00 EXTHDD

$ touch /run/media/marcos/EXTHDD/test.txt
touch: cannot touch '/run/media/marcos/EXTHDD/test.txt': Permission denied

$ findmnt /run/media/marcos/EXTHDD 
TARGET                   SOURCE   FSTYPE OPTIONS
/run/media/marcos/EXTHDD /dev/sdb ext3   rw,nosuid,nodev,relatime,data=ordered
```

(there is nothing related to this device in fstab)

With Windows filesystems (FAT, exFAT, NTFS) it works fine:

```
$ ls -l /run/media/marcos/
total 4
drwxr-xr-x 3 marcos marcos 4096 May 15 14:20 9047-C0D3

$ touch /run/media/marcos/9047-C0D3/test.txt

$ ls -l /run/media/marcos/9047-C0D3/
total 4
drwxr-xr-x 2 marcos marcos 4096 May 20  2016 'System Volume Information'
-rw-r--r-- 1 marcos marcos    0 May 15 14:20  test.txt

$ findmnt /run/media/marcos/9047-C0D3
TARGET                      SOURCE   FSTYPE OPTIONS
/run/media/marcos/9047-C0D3 /dev/sdc vfat   rw,nosuid,nodev,relatime,uid=1000,gid=1000,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,showexec,utf8,flush,errors=remount-ro
```

(9047-C0D3 is a FAT32 USB Flash Disk)

So...

Let's try the obvious:

```
# chown marcos:marcos /run/media/marcos/EXTHDD/

$ ls -l /run/media/marcos/
total 4
drwxr-xr-x 3 marcos marcos 4096 May 15 14:30 EXTHDD

$ touch /run/media/marcos/EXTHDD/test.txt

$ ls -l /run/media/marcos/EXTHDD/
total 16
drwx------ 2 root   root   16384 May 15 14:00 lost+found
-rw-r--r-- 1 marcos marcos     0 May 15 14:30 test.txt
```

Is this a bug?

**Comment by marcosfrm** [comment]
I think the FAT/exFAT/NTFS case works because the driver "upgrade" the mount point to especified `uig=` and `gid=`:

```
# mkdir /mnt/foo

# ls -l /mnt
total 0
drwxr-xr-x 2 root root 40 May 18 11:51 foo

# mount -t vfat /dev/sdb /mnt/foo -o uid=1000,gid=1000

# ls -l /mnt
total 4
drwxr-xr-x 3 marcos marcos 4096 Jan  1  1970 foo
             ^^^^^^ ^^^^^^
```

For native filesystems it does not occur. Does mount(8) have any generic option to set mount point owner?

Is `chown(mount_point_to_use, uid, gid);` after successfully mount an option (at least for volumes not listed in fstab)?

**Comment by vojtechtrefny** [comment]
I'll take a look at this.

The `uid` and `gid` options are specific for fat and ntfs (and few other) filesystems, ext doesn't support them.

**Comment by marcosfrm** [comment]
Thanks.

Btw, `mount_point_to_use` for non-fstab mounts is created with 0700 permissions:

https://github.com/storaged-project/udisks/blob/master/src/udiskslinuxfilesystem.c#L1476

But after mount it will be 0755 anyway. On #L1301 (fstab mounts) it is created with 0755.

Not that it matters much... just nitpicking. :-)

**Comment by pothos** [comment]
What about the take-ownership option when the filesystem is created through UDisks? For me that is the solution. In GNOME Disks this is only used for NTFS/FAT by default, but in e.g. d-feet the Format call on the block interface of a device takes these arguments: 'ext4', {'take-ownership': GLib.Variant('b', True)}

**Comment by marcosfrm** [comment]
The ownership is saved in filesystem itself right? So *all* fresh new Linux native filesystems are owned by root, be it created by GNOME Disks or not. This behaviour is very unfriendly for users.

But changing ownership of random filesystems may not be a goot idea. @pothos suggestion it much better.

**Comment by pothos** [comment]
Difficult to judge what's right. It depens on the scenario, if a SD card with a Raspberry Pi rootfs is mounted then the system files there need to be manipulated and created as uid 0 which also happens to be the local root user. Chaning the ownership of the '/' directory is not what I would expect to happen during a mount which then affects this rootfs when being used.

**Comment by vojtechtrefny** [comment]
@pothos  Thanks, I didn't know that we have something like "take-ownership". It is definitely the best solution here -- GNOME Disks should really use this option for all filesystems that support ownership.

@vpodzime also suggested we could create a new function `Filesystem.TakeOwnership()` to change ownership of a mountpoint (filesystem) -- this should be a more "user-friendly" to solve this problem for filesystems created with other tools than GNOME Disks.

**Comment by pothos** [comment]
That's a good idea! Yes, take-ownership should be made accessible for ext4 in the format dialog somehow.

**Comment by vpodzime** [comment]
> In GNOME Disks this is only used for NTFS/FAT by default...

Hmm, what is reason to use the option for these two file systems that don't support ownership and access control? These are **mounted** with *uid* and *gid* options which handle this stuff, right?

**Comment by pothos** [comment]
True! It seems to make no difference… :D

**Comment by vpodzime** [comment]
I believe this can be closed now that #350 is merged.
